### PR TITLE
[reward, recipe] fix: stop mapping HTTP scorer errors to 0 rewards

### PR DESCRIPTION
**Pull request by ShuoleiWang**
### What does this PR do?

The generic HTTP scorer client previously returned `{"score": 0.0}` in three failure cases:

- the server returned a non-200 HTTP status;
- the response contained an `error`;
- the response contained an empty `scores` list.

A real reward can also be `0.0`. This means the reward manager could not tell the difference between a valid zero reward and a failed scorer request.

The root cause was that the HTTP client returned a normal score instead of raising an exception. Because no exception was raised, the reward manager could not apply its existing `required` policy.

This PR changes the behavior to:

```text
Before:
HTTP/service failure -> score 0 -> training continues

After:
HTTP/service failure -> exception -> reward manager applies required policy
```

The OCR HTTP scorer recipe now sets `required=true`, so these failures stop the training run instead of contributing a zero reward.

A real response such as `{"scores": [0.0]}` remains valid.

The existing behavior for optional scorers is unchanged: when `required=false`, the reward manager may still replace a failed scorer result with zero and continue.

Related work:

- [#116](https://github.com/verl-project/verl-omni/pull/116) introduced the generic HTTP scorer client.
- [#289](https://github.com/verl-project/verl-omni/pull/289) introduced the existing `required` policy and a separate latent HTTP scorer. This PR applies the failure behavior to the older generic HTTP scorer client.

### Checklist Before Starting

- [x] Searched [open PRs for HTTP scorer error handling](https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+is%3Aopen+http+scorer). No matching open PR was found.
- [x] The PR title follows `[{modules}] {type}: {description}`.

### Test

CPU tests:

```bash
pytest -q \
  tests/utils/reward_score/test_http_scorer_client_on_cpu.py \
  tests/utils/reward_score/test_latent_http_scorer_client_on_cpu.py
```

Result:

```text
12 passed
```

Static checks:

```bash
ruff check \
  verl_omni/utils/reward_score/http_scorer_client.py \
  tests/utils/reward_score/test_http_scorer_client_on_cpu.py

ruff format --check \
  verl_omni/utils/reward_score/http_scorer_client.py \
  tests/utils/reward_score/test_http_scorer_client_on_cpu.py

bash -n examples/flowgrpo_trainer/qwen_image/run_qwen_image_ocr_reward_server.sh
git diff --check
```

All checks passed. No GPU is required for this change.

### API and Usage Example

There is no function-signature or configuration-schema change.

Direct callers of `compute_score()` now receive a `RuntimeError` for non-200 responses, service-reported errors, and empty score responses.

The OCR recipe now enables the existing required-reward behavior:

```bash
'+reward.reward_functions.ocr.required=true'
```

### Design & Code Changes

- Update the generic HTTP scorer client to raise exceptions instead of returning zero for the three failure cases.
- Enable `required=true` in the OCR HTTP scorer recipe.
- Document how `required` handles scorer failures.
- Add CPU tests covering:
  - a valid zero reward;
  - a non-200 HTTP response;
  - a service-reported error;
  - an empty score response.

Retries, timeout policy, and non-finite score validation are intentionally outside the scope of this PR.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Applied the required pre-commit checks.
- [x] Updated the documentation.
- [x] Added CPU unit tests covering the changed behavior.

### AI Assistance

AI assistance was used for repository inspection, implementation drafting, tests, and preparation of this PR description. I reviewed every changed line and ran the checks listed above.

### CODE REVIEW

**Comment by zhtmike** [comment]
I think the training loop should be robust enough to handling the occasionly error from reward server (e.g., occasionaly internet issue, reward inference server is fully busy). So training should be continue with this kind of error.

So is this a really good fix?

**Comment by ShuoleiWang** [comment]
> I think the training loop should be robust enough to handling the occasionly error from reward server (e.g., occasionaly internet issue, reward inference server is fully busy). So training should be continue with this kind of error.
> 
> So is this a really good fix?

@zhtmike You're right, the reward-server failure should not stop training immediately. I ignored the generic HTTP scorer currently has no retry mechanism.

This PR is trying to address: **converting a scorer failure directly into `0` makes it indistinguishable from a valid zero reward, which can distort the training signal.**

Do you think we should do something for this issue?

**Comment by zhtmike** [comment]
> > I think the training loop should be robust enough to handling the occasionly error from reward server (e.g., occasionaly internet issue, reward inference server is fully busy). So training should be continue with this kind of error.
> > So is this a really good fix?
> 
> @zhtmike You're right, the reward-server failure should not stop training immediately. I ignored the generic HTTP scorer currently has no retry mechanism.
> 
> This PR is trying to address: **converting a scorer failure directly into `0` makes it indistinguishable from a valid zero reward, which can distort the training signal.**
> 
> Do you think we should do something for this issue?

as you said, can we implement a retry loop for remote http server? it should be helpful -> Tolerate the occasional failure, but can spot the systematic failure.

And can we run a e2e after implementing this, update some figures to prove that it does work?

**Comment by ShuoleiWang** [comment]
> as you said, can we implement a retry loop for remote http server? it should be helpful -> Tolerate the occasional failure, but can spot the systematic failure.
> 
> And can we run a e2e after implementing this, update some figures to prove that it does work?

OK, very insightful!

**Comment by ShuoleiWang** [comment]
> as you said, can we implement a retry loop for remote http server? it should be helpful -> Tolerate the occasional failure, but can spot the systematic failure.
> 
> And can we run a e2e after implementing this, update some figures to prove that it does work?

@zhtmike Hi, according to your feedback, I have added bounded retries to the generic HTTP scorer client for **transport failures, timeouts, and HTTP 408/429/5xx responses**. The default is 2 retries after the initial request, with 0.5s/1.0s backoff. The client still fails immediately on non-retryable responses.

I reused the **existing Qwen-Image FlowGRPO E2E setup on 4 H20 GPUs**. Its train batch size of 8 and `rollout.n=2` produced 16 scoring inputs for one step. I used the complete Qwen/Qwen-Image checkpoint and real OCR prompts from the Flow-GRPO training dataset. Each scoring input contained one generated JPEG and its OCR prompt, and full body/JPEG hashes confirmed that the request stayed unchanged across retries.

For the transient-failure test, I placed a local fault-injection proxy in front of PaddleOCR. The proxy returned HTTP 503 for the first two attempts of each scoring input, while the HTTP client waited approximately 0.5s/1.0s and retried. On the 3rd attempt, the proxy forwarded the unchanged request body to PaddleOCR. All 16 upstream calls returned finite scores, and training completed global step 1. For the persistent-failure test, the proxy always returned 503 and never called PaddleOCR. No request exceeded 3 attempts, and training stopped before a global step with `Required sub-reward ... failed after 3 attempts`. I also added a CPU integration test with a real local HTTP server, covering both retry recovery and retry exhaustion (1 passed).

**Comment by zhtmike** [comment]
Could you provide a reward curve in jpeg format?

File: examples/flowgrpo_trainer/qwen_image/run_qwen_image_ocr_reward_server.sh [hunk 1]
```diff
@@ -83,6 +83,7 @@ python3 -m verl_omni.trainer.main_diffusion \
     "+reward.reward_functions.ocr.path=pkg://verl_omni.utils.reward_score.http_scorer_client" \
     '+reward.reward_functions.ocr.name=compute_score' \
     '+reward.reward_functions.ocr.weight=1.0' \
+    '+reward.reward_functions.ocr.required=true' \
```
Comment (knlnguyen1802) [inline examples/flowgrpo_trainer/qwen_image/run_qwen_image_ocr_reward_server.sh]: Already have too many script under /flowgpo/qwen_image/ consider mention this in docs only

**Comment by ShuoleiWang** [comment]
> Could you provide a reward curve in jpeg format?
pr366_qwen_image_reward_curve


I re-ran the experiment and here is the reward curve in JPEG format. The upper panel shows the raw mean OCR reward per step; the lower panel shows the number of injected 503 responses per step.
I reused the repository’s existing Qwen-Image FlowGRPO E2E setup and ran 2 independent 20-step traces on each 4 H20 GPUs. In the **sparse-fault run**, a local proxy injected 1 or 2 HTTP 503 responses into a small subset of scoring requests. All requests completed successfully within the configured three-attempt limit, and training reached step 20.
The 2 reward curves are not expected to match exactly the same, because the diffusion rollouts are stochastic and generated different images.

File: examples/flowgrpo_trainer/qwen_image/run_qwen_image_ocr_reward_server.sh (hunk 1 above)
Comment (ShuoleiWang) [reply examples/flowgrpo_trainer/qwen_image/run_qwen_image_ocr_reward_server.sh]: Done

**Comment by zhtmike** [comment]
> > Could you provide a reward curve in jpeg format?
> 
> pr366_qwen_image_reward_curve
> I re-ran the experiment and here is the reward curve in JPEG format. The upper panel shows the raw mean OCR reward per step; the lower panel shows the number of injected 503 responses per step. I reused the repository’s existing Qwen-Image FlowGRPO E2E setup and ran 2 independent 20-step traces on each 4 H20 GPUs. In the **sparse-fault run**, a local proxy injected 1 or 2 HTTP 503 responses into a small subset of scoring requests. All requests completed successfully within the configured three-attempt limit, and training reached step 20. The 2 reward curves are not expected to match exactly the same, because the diffusion rollouts are stochastic and generated different images.

Thanks!

File: docs/start/http_scorer.md [hunk 2]
```diff
@@ -74,7 +74,10 @@ python3 -m verl_omni.trainer.main_diffusion \
     "+reward.reward_functions.my_reward.path=pkg://verl_omni.utils.reward_score.http_scorer_client" \
     '+reward.reward_functions.my_reward.name=compute_score' \
     '+reward.reward_functions.my_reward.weight=1.0' \
+    '+reward.reward_functions.my_reward.required=true' \
```
Comment (zhtmike) [inline docs/start/http_scorer.md:77]: `required` <--- this name is too ambiguous, pls use a better name

Comment (ShuoleiWang) [reply docs/start/http_scorer.md:77]: The `required` tag is not introduced by this PR, it is an existing MultiVisualRewardManager option introduced in #289 and already used by the SD3.5 DRM recipe. And it means a sub-reward error is propagated instead of being converted to a 0 score. I'm worried renaming it may damage other features or configurations.

Comment (zhtmike) [reply docs/start/http_scorer.md:77]: I see, let us keep it

Comment (ShuoleiWang) [reply docs/start/http_scorer.md:77]: It seems the CPU failure in CI was caused by the Qwen3-Omni test incompatibility, and it has already been fixed on main. I have synced the latest main. Could you please retrigger the ci label?

**Review by zhtmike — APPROVED** [review]

**Comment by zhtmike** [comment]
thanks for the pr！
